crates/vault-registry/src/migration.rs

```diff
@@ -0,0 +1,320 @@
+use super::*;
+use frame_support::{pallet_prelude::StorageVersion, traits::OnRuntimeUpgrade};
+
+#[cfg(feature = "try-runtime")]
+use sp_std::vec::Vec;
+
+/// The log target.
+const TARGET: &'static str = "runtime::vault_registry::migration";
+
+pub mod vault_capacity {
+    use super::*;
+    use frame_support::pallet_prelude::*;
+
+    type SignedFixedPoint<T> = <T as currency::Config>::SignedFixedPoint;
+
+    fn clear_reward_storage<T: Config>(mut weight: Weight, item: &str) {
+        let res = frame_support::migration::clear_storage_prefix(b"VaultRewards", item.as_bytes(), b"", None, None);
+        weight.saturating_accrue(T::DbWeight::get().writes(res.backend.into()));
+
+        log::info!(
+            target: TARGET,
+            "Cleared '{}' entries from '{item}' storage prefix",
+            res.unique
+        );
+
+        if res.maybe_cursor.is_some() {
+            log::error!(target: TARGET, "Storage prefix '{item}' is not completely cleared");
+        }
+    }
+
+    #[derive(Debug, Encode, Decode)]
+    struct RewardsState<SignedFixedPoint> {
+        stake_entries: u32,
+        total_rewards_native: SignedFixedPoint,
+        total_rewards_wrapped: SignedFixedPoint,
+    }
+
+    pub struct RewardsMigration<Runtime, VaultRewardsInstance>(
+        sp_std::marker::PhantomData<(Runtime, VaultRewardsInstance)>,
+    );
+
+    impl<
+            Runtime: Config
+                + reward::Config<
+                    VaultRewardsInstance,
+                    StakeId = DefaultVaultId<Runtime>,
+                    CurrencyId = CurrencyId<Runtime>,
+                    SignedFixedPoint = SignedFixedPoint<Runtime>,
+                > + staking::Config<CurrencyId = CurrencyId<Runtime>, SignedFixedPoint = SignedFixedPoint<Runtime>>,
+            VaultRewardsInstance: 'static,
+        > OnRuntimeUpgrade for RewardsMigration<Runtime, VaultRewardsInstance>
+    {
+        #[cfg(feature = "try-runtime")]
+        fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+            let prev_count = reward::migration::v0::Stake::<Runtime, VaultRewardsInstance>::iter().count();
+            log::info!(target: TARGET, "{} stake entries will be migrated", prev_count);
```

Comment:
Instead of comparing the totals, can we compare the balances of each vault after using `with_transaction` to do a dry-run of `withdraw_rewards`?